wagtail/models/__init__.py

```diff
@@ -4606,3 +4613,188 @@ class Meta:
         unique_together = [
             ("page", "user"),
         ]
+
+
+class ReferenceIndex(models.Model):
+    """
+    Records references between objects for quick retrieval of object usage.
+
+    References are extracted from Foreign Keys, Chooser Blocks in StreamFields, and links in Rich Text Fields.
+    This index allows us to efficiently find all of the references to a particular object from all of these sources.
+    """
+
+    # The object where the reference was extracted from
+
+    # content_type represents the content type of the model that contains
+    # the field where the reference came from. If the model sub-classes another
+    # concrete model (such as Page), that concrete model will be set in
+    # base_content_type, otherwise it would be the same as content_type
+    content_type = models.ForeignKey(
+        ContentType, on_delete=models.CASCADE, related_name="+"
+    )
+    base_content_type = models.ForeignKey(
+        ContentType, on_delete=models.CASCADE, related_name="+"
+    )
+    object_id = models.CharField(
+        max_length=255,
+        verbose_name=_("object id"),
+    )
+
+    # The object that has been referenced
+    # to_content_type is always the base content type of the referenced object
+    to_content_type = models.ForeignKey(
+        ContentType, on_delete=models.CASCADE, related_name="+"
+    )
+    to_object_id = models.CharField(
+        max_length=255,
+        verbose_name=_("object id"),
+    )
+
+    # The path to the field on content_type where the reference was extracted from
+    # This is a dotted path, always starting with a field/child relation name. If
+    # the reference was extracted from an inline panel or streamfield, other components
+    # of the path can be used to locate where the reference was extracted.
+    #
+    # For example, say we have a StreamField called 'body' which has a struct block type
+    # called 'my_struct_block' that has a field called 'my_field'. If we extracted a
+    # reference from that field, the path would be set to the following:
+    #
+    # 'body.my_struct_block.my_field'
+    #
+    # We can then use this with the 'content_type' to find the original definition of
+    # the field block and display information to the user about where the reference was
+    # extracted from.
+    path = models.TextField()
```

Comment:
Probably split this into two separate path fields:

 - `model_path` (For example `body.my_block.my_field`) - Used to find details of the field the reference came from, without having to retrieve the original object
 - `content_path (For example `body.bdc70d8b-e7a2-4c2a-bf43-2a3e3fcbbe86.my_field`) - Used to permalink the user to the specific block/field that contains the reference